site/app/libraries/Access.php

```diff
@@ -0,0 +1,170 @@
+        $this->permissions["grading.save_general_comment"] = self::ALLOW_MIN_STUDENT | self::CHECK_GRADEABLE_MIN_GROUP | self::CHECK_GRADING_SECTION_GRADER | self::CHECK_PEER_ASSIGNMENT_STUDENT | self::CHECK_HAS_SUBMISSION;
+        $this->permissions["grading.get_mark_data"] = self::ALLOW_MIN_STUDENT | self::CHECK_GRADEABLE_MIN_GROUP | self::CHECK_GRADING_SECTION_GRADER | self::CHECK_PEER_ASSIGNMENT_STUDENT;
+        $this->permissions["grading.get_gradeable_comment"] = self::ALLOW_MIN_STUDENT | self::CHECK_GRADEABLE_MIN_GROUP | self::CHECK_GRADING_SECTION_GRADER | self::CHECK_PEER_ASSIGNMENT_STUDENT;
+        $this->permissions["grading.add_one_new_mark"] = self::CHECK_GRADEABLE_MIN_GROUP | self::CHECK_GRADING_SECTION_GRADER;
+        $this->permissions["grading.delete_one_mark"] = self::CHECK_GRADEABLE_MIN_GROUP | self::CHECK_GRADING_SECTION_GRADER;
+        $this->permissions["grading.import_teams"] = self::ALLOW_MIN_INSTRUCTOR | self::CHECK_CSRF;
+        $this->permissions["grading.export_teams"] = self::ALLOW_MIN_INSTRUCTOR;
+        $this->permissions["grading.submit_team_form"] = self::ALLOW_MIN_INSTRUCTOR;
+        $this->permissions["grading.verify_grader"] = self::ALLOW_MIN_TA;
+        $this->permissions["grading.verify_all"] = self::ALLOW_MIN_TA;
+    }
+
+    /**
+     * Check if the currently logged in user is allowed to do an action
+     * @param string $action Name of the action (see Access::$permissions)
+     * @param array $args Any extra arguments that are required to check permissions
+     * @return bool True if they are allowed to do that action
+     */
+    public function canI(string $action, $args = []) {
+        if (!array_key_exists($action, $this->permissions)) {
+            return false;
+        }
+        $checks = $this->permissions[$action];
+
+        //Some things may be available when there is no user
+        $user = $this->core->getUser();
+        if ($user === null) {
+            return !!($checks & self::ALLOW_LOGGED_OUT);
+        }
+        //Check user group first
+        $group = $user->getGroup();
+        if ($group === self::USER_GROUP_STUDENT && !($checks & self::ALLOW_STUDENT)) {
+            return false;
+        } else if ($group === self::USER_GROUP_GRADER && !($checks & self::ALLOW_GRADER)) {
+            return false;
+        } else if ($group === self::USER_GROUP_TA && !($checks & self::ALLOW_TA)) {
+            return false;
+        } else if ($group === self::USER_GROUP_INSTRUCTOR && !($checks & self::ALLOW_INSTRUCTOR)) {
+            return false;
+        }
+
+        if ($checks & self::CHECK_CSRF) {
+            if ($this->core->checkCsrfToken($_POST['csrf_token'])) {
+                return false;
+            }
+        }
+
+        /* @var Gradeable|null $gradeable */
+        $gradeable = $args["gradeable"] ?? null;
```

Comment:
This function is to be only used with gradeable endpoints?

Reply:
Only gradeable-related checks will need it. The alternative is copying that line into every gradeable-related test, which I'm completely fine with.

Reply:
You still have a potential for throwing a null error for the grading tests. But the intention is to expand this beyond just testing of the gradeable object?